File: py/processor/Processor/commands.py
from .core import process_task


async def process_chart_arguments(arguments, platforms, tickerId=None, defaults={}):
	payload = await process_task(
		{
			"arguments": arguments,
			"platforms": platforms,
			"tickerId": tickerId,
			"defaults": defaults
		},
		"parser",
		endpoint="/chart"
	)
	return payload.get("message"), payload.get("response")

async def process_heatmap_arguments(arguments, platforms):
	payload = await process_task(
		{
			"arguments": arguments,
			"platforms": platforms
		},
		"parser",
		endpoint="/heatmap"
	)
	return payload.get("message"), payload.get("response")

async def process_quote_arguments(arguments, platforms, tickerId=None):
	payload = await process_task(
		{
			"arguments": arguments,
			"platforms": platforms,
			"tickerId": tickerId
		},
		"parser",
		endpoint="/quote"
	)
	return payload.get("message"), payload.get("response")

async def process_detail_arguments(arguments, platforms, tickerId=None):
	payload = await process_task(
		{
			"arguments": arguments,
			"platforms": platforms,
			"tickerId": tickerId
		},
		"parser",
		endpoint="/detail"
	)
	return payload.get("message"), payload.get("response")

File: rs/processor/src/commands.rs
use serde_json::{Value, json};

use crate::process_task;

pub async fn process_chart_arguments(arguments: Vec<&str>, platforms: Vec<&str>, ticker_id: Option<&str>, defaults: Value) -> (Value, Value) {
	let payload = process_task(
		json!({
			"arguments": arguments,
			"platforms": platforms,
			"tickerId": ticker_id,
			"defaults": defaults
		}),
		"parser",
		Some("/chart"),
		None,
		None,
		None
	).await;

	if let Err(e) = payload {
		(Value::String(e.to_string()), Value::Null)
	} else {
		let payload = unsafe { payload.unwrap_unchecked() };
		(payload.get("message").unwrap().clone(), payload.get("response").unwrap().clone())
	}
}

pub async fn process_heatmap_arguments(arguments: Vec<&str>, platforms: Vec<&str>) -> (Value, Value) {
	let payload = process_task(
		json!({
			"arguments": arguments,
			"platforms": platforms
		}),
		"parser",
		Some("/heatmap"),
		None,
		None,
		None
	).await;

	if let Err(e) = payload {
		(Value::String(e.to_string()), Value::Null)
	} else {
		let payload = unsafe { payload.unwrap_unchecked() };
		(payload.get("message").unwrap().clone(), payload.get("response").unwrap().clone())
	}
}

pub async fn process_quote_arguments(arguments: Vec<&str>, platforms: Vec<&str>, ticker_id: Option<&str>) -> (Value, Value) {
	let payload = process_task(
		json!({
			"arguments": arguments,
			"platforms": platforms,
			"tickerId": ticker_id
		}),
		"parser",
		Some("/quote"),
		None,
		None,
		None
	).await;

	if let Err(e) = payload {
		(Value::String(e.to_string()), Value::Null)
	} else {
		let payload = unsafe { payload.unwrap_unchecked() };
		(payload.get("message").unwrap().clone(), payload.get("response").unwrap().clone())
	}
}

pub async fn process_detail_arguments(arguments: Vec<&str>, platforms: Vec<&str>, ticker_id: Option<&str>) -> (Value, Value) {
	let payload = process_task(
		json!({
			"arguments": arguments,
			"platforms": platforms,
			"tickerId": ticker_id
		}),
		"parser",
		Some("/detail"),
		None,
		None,
		None
	).await;

	if let Err(e) = payload {
		(Value::String(e.to_string()), Value::Null)
	} else {
		let payload = unsafe { payload.unwrap_unchecked() };
		(payload.get("message").unwrap().clone(), payload.get("response").unwrap().clone())
	}
}